include/fiction/algorithms/network_transformation/inverter_balancing.hpp

```diff
@@ -0,0 +1,320 @@
+//
+// Created by benjamin on 13.04.23.
+//
+
+#ifndef FICTION_INVERTER_BALANCING_HPP
+#define FICTION_INVERTER_BALANCING_HPP
+
+#include "fiction/traits.hpp"
+#include "fiction/types.hpp"
+#include "fiction/utils/name_utils.hpp"
+
+#include <mockturtle/traits.hpp>
+#include <mockturtle/utils/node_map.hpp>
+#include <mockturtle/views/fanout_view.hpp>
+#include <mockturtle/views/topo_view.hpp>
+
+#include <type_traits>
+
+#if (PROGRESS_BARS)
+#include <mockturtle/utils/progress_bar.hpp>
+#endif
+
+namespace fiction
+{
+
+namespace detail
+{
+
+template <typename Ntk>
+class inverter_balancing_impl
+{
+  public:
+    explicit inverter_balancing_impl(const Ntk& ntk_src) : ntk{ntk_src} {}
+
+    Ntk run()
+    {
+        // compute maximum sizes for vectors
+        std::uint64_t vec_size{0};
+        fo_ntk.foreach_node(
+            [this, &vec_size](const auto& n)
+            {
+                if (fo_ntk.is_fanout(n) && ntk.fanout_size(n) >= 2)
+                {
+                    ++vec_size;
+                }
+            });
+        x_inv.reserve(vec_size);
+        m_inv.reserve(vec_size);
+        blc_fos.reserve(vec_size);
+
+        auto  init     = mockturtle::initialize_copy_network<Ntk>(ntk);
+        auto& ntk_dest = init.first;
+        auto& old2new  = init.second;
+
+        const auto gather_fanin_signals = [this, &ntk_dest, &old2new](const auto& n)
+        {
+            std::vector<typename Ntk::signal> children{};
+
+            if (fo_ntk.is_fanout(n))
+            {
+                auto fanout_inv = fanouts(fo_ntk, n);
+                auto balance    = std::all_of(fanout_inv.begin(), fanout_inv.end(),
+                                              [this](const auto& fo_node) { return fo_ntk.is_inv(fo_node); });
+                if (balance && fanout_inv.size() > 1)
+                {
+                    x_inv.emplace_back(fanout_inv[1]);
+                    blc_fos.emplace_back(n);
+                    m_inv.emplace_back(fanout_inv[0]);
+                }
+            }
+
+            ntk.foreach_fanin(n,
+                              [this, &old2new, &children, &n](const auto& f)
+                              {
+                                  auto fn = ntk.get_node(f);
+
+                                  // New children for nodes with deleted Inverters as fan-in
+                                  for (int i = 0; i < blc_fos.size(); ++i)
```

Comment:
Loop variables that iterate over sizes should be of an unsigned data type. Ideally the same as the size type. For instance `std::size_t`